docs/hub/how-to-upstream.md

```diff
@@ -0,0 +1,257 @@
+# How to integrate upstream methods in your library
+
+Now that you've seen more downstream methods for downloading files, it is time to introduce you to additional *upstream* methods. Upstream methods refers to methods that publish files to the Hub from your library. This guide will show you how to:
+
+* Use the `HfApi` class to manage a repository.
+* Use the `Repository` class to handle files and version control a repository with Git-like commands.
+
+## `HfApi`
+
+The `HfApi` class is a high level class that wraps around HTTP requests. There are many useful tasks you can accomplish with the `HfApi` class including: 
+
+- List and filter models.
+- Inspect model or dataset metadata.
+- Delete a repository.
+- Change the visibility of a repository.
+
+### List and filter
+
+It can be helpful for users to see a list of available models, and filter them according to a specific language or framework. This can be especially useful for library and organization owners who want to view all the models they own. Use the `list_models` function with the `filter` parameter to search for a model of interest:
+
+```python
+>>> from huggingface_hub import HfApi
+>>> api = HfApi()
+
+# List all models.
+>>> api.list_models()
+
+# List only text classification models.
+>>> api.list_models(filter="text-classification")
+
+# List only Russian models compatible with PyTorch.
+>>> api.list_models(filter=("ru", "pytorch"))
+
+# List only the models trained on the "common_voice" dataset.
+>>> api.list_models(filter="dataset:common_voice")
+
+# List only the models from the spaCy library.
+>>> api.list_models(filter="spacy")
+```
+
+Explore available public datasets with `list_datasets`:
+
+```python
+>>> from huggingface_hub import HfApi
+>>> api = HfApi()
+
+# List only text classification datasets.
+>>> api.list_datasets(filter="task_categories:text-classification")
+
+# List only datasets in Russian for language modeling.
+>>> api.list_datasets(filter=("languages:ru", "task_ids:language-modeling"))
+```
+
+### Inspect model or dataset metadata
+
+Get important information about a model or dataset as shown in the following:
+
+```python
+>>> from huggingface_hub import HfApi
+>>> api = HfApi()
+
+# Get metadata of a single model.
+>>> api.model_info("distilbert-base-uncased")
+
+# Get metadata of a single dataset.
+>>> api.dataset_info("glue")
+```
+
+### Delete a repository
+
+Delete a repository with `delete_repo`. Make sure you are certain you want to delete a repository because this is an irreversible process!
+
+1. Get your Hugging Face API token:
+
+   ```python
+   >>> from huggingface_hub import HfFolder
+   >>> folder = HfFolder()
+   >>> token = folder.get_token()
+   ```
+
+2. Pass your token and the name of the repository to `delete_repo`:
+
+   ```python
+   >>> api.delete_repo(token=YOUR_HF_API_TOKEN, name=REPO_NAME)
+   ```
+
+Delete a dataset repository by adding the `type` parameter:
```

Comment:
Should we mention what the most common types are?

Reply:
If the only types are `dataset` and `spaces` as shown [here](https://github.com/huggingface/huggingface_hub/blob/e665982eacf9b8b05115d324ba1202c7d542fcbf/src/huggingface_hub/constants.py#L34), then I don't think we need to mention it since Spaces isn't quite open to everyone yet :)